File: sdk/go/record_service.go
package cratebase

import "strconv"

// RecordService provides CRUD + auth operations scoped to one collection.
// Auth-only methods (AuthWithPassword, AuthRefresh, ...) only make sense
// for auth-typed collections but are exposed unconditionally since the
// client has no way to know a collection's type ahead of a request.
type RecordService struct {
	client             *Client
	collectionIdOrName string
}

func (s *RecordService) basePath() string {
	return "/api/collections/" + s.collectionIdOrName + "/records"
}

func (s *RecordService) authPath(action string) string {
	return "/api/collections/" + s.collectionIdOrName + "/" + action
}

// GetList fetches one page of records. page defaults to 1, perPage to 30
// when 0.
func (s *RecordService) GetList(page, perPage int, options ListOptions) (*ListResult[RecordModel], error) {
	if page == 0 {
		page = 1
	}
	if perPage == 0 {
		perPage = 30
	}
	var result ListResult[RecordModel]
	err := s.client.Send(s.basePath(), SendOptions{
		Method: "GET",
		Query: map[string]string{
			"page":    strconv.Itoa(page),
			"perPage": strconv.Itoa(perPage),
			"filter":  options.Filter,
			"sort":    options.Sort,
		},
	}, &result)
	if err != nil {
		return nil, err
	}
	return &result, nil
}

// GetFullList fetches every page and concatenates the results. Convenient
// for small collections; prefer GetList with pagination for large ones.
// batchSize defaults to 200 when 0.
func (s *RecordService) GetFullList(options ListOptions, batchSize int) ([]RecordModel, error) {
	if batchSize == 0 {
		batchSize = 200
	}
	var items []RecordModel
	page := 1
	for {
		result, err := s.GetList(page, batchSize, options)
		if err != nil {
			return nil, err
		}
		items = append(items, result.Items...)
		if page >= result.TotalPages {
			break
		}
		page++
	}
	return items, nil
}

// GetOne fetches a single record by id.
func (s *RecordService) GetOne(id string) (RecordModel, error) {
	var record RecordModel
	err := s.client.Send(s.basePath()+"/"+id, SendOptions{Method: "GET"}, &record)
	return record, err
}

// Create creates a record. Pass a map[string]any/struct for a JSON body,
// or a *MultipartForm for collections with file fields.
func (s *RecordService) Create(data any) (RecordModel, error) {
	var record RecordModel
	err := s.client.Send(s.basePath(), SendOptions{Method: "POST", Body: data}, &record)
	return record, err
}

// Update partially updates a record. Pass a map[string]any/struct for a
// JSON body, or a *MultipartForm for collections with file fields.
func (s *RecordService) Update(id string, data any) (RecordModel, error) {
	var record RecordModel
	err := s.client.Send(s.basePath()+"/"+id, SendOptions{Method: "PATCH", Body: data}, &record)
	return record, err
}

// Delete deletes a record by id.
func (s *RecordService) Delete(id string) error {
	return s.client.Send(s.basePath()+"/"+id, SendOptions{Method: "DELETE"}, nil)
}

// AuthWithPassword logs in with identity (whatever the collection's
// configured identity field is — email by default) and password.
func (s *RecordService) AuthWithPassword(identity, password string) (*AuthResponse[RecordModel], error) {
	var result AuthResponse[RecordModel]
	err := s.client.Send(s.authPath("auth-with-password"), SendOptions{
		Method: "POST",
		Body:   map[string]any{"identity": identity, "password": password},
	}, &result)
	if err != nil {
		return nil, err
	}
	s.client.AuthStore.Save(result.Token, result.Record)
	return &result, nil
}

// AuthRefresh renews the current session, returning a fresh token/record.
func (s *RecordService) AuthRefresh() (*AuthResponse[RecordModel], error) {
	var result AuthResponse[RecordModel]
	err := s.client.Send(s.authPath("auth-refresh"), SendOptions{Method: "POST"}, &result)
	if err != nil {
		return nil, err
	}
	s.client.AuthStore.Save(result.Token, result.Record)
	return &result, nil
}

// ListAuthMethods lists available auth methods. Pass your app's OAuth2
// redirect URI/deep link and each provider's AuthURL comes back ready to
// open directly.
func (s *RecordService) ListAuthMethods(redirectUri string) (*AuthMethodsResponse, error) {
	var result AuthMethodsResponse
	err := s.client.Send(s.authPath("auth-methods"), SendOptions{
		Method: "GET",
		Query:  map[string]string{"redirectUri": redirectUri},
	}, &result)
	if err != nil {
		return nil, err
	}
	return &result, nil
}

// AuthWithOAuth2 completes an OAuth2 login: exchange the code your app
// received at redirectUri (the same one used to build the authUrl from
// ListAuthMethods) for a session.
func (s *RecordService) AuthWithOAuth2(provider, code, redirectUri string) (*AuthResponse[RecordModel], error) {
	var result AuthResponse[RecordModel]
	err := s.client.Send(s.authPath("auth-with-oauth2"), SendOptions{
		Method: "POST",
		Body:   map[string]any{"provider": provider, "code": code, "redirectUri": redirectUri},
	}, &result)
	if err != nil {
		return nil, err
	}
	s.client.AuthStore.Save(result.Token, result.Record)
	return &result, nil
}

// RequestVerification sends a verification email if email matches an
// account — always resolves regardless, so it can't be used to enumerate
// accounts.
func (s *RecordService) RequestVerification(email string) error {
	return s.client.Send(s.authPath("request-verification"), SendOptions{
		Method: "POST",
		Body:   map[string]any{"email": email},
	}, nil)
}

// ConfirmVerification confirms a verification token from the emailed link.
func (s *RecordService) ConfirmVerification(token string) error {
	return s.client.Send(s.authPath("confirm-verification"), SendOptions{
		Method: "POST",
		Body:   map[string]any{"token": token},
	}, nil)
}

// RequestPasswordReset sends a password reset email if email matches an
// account — always resolves regardless, so it can't be used to enumerate
// accounts.
func (s *RecordService) RequestPasswordReset(email string) error {
	return s.client.Send(s.authPath("request-password-reset"), SendOptions{
		Method: "POST",
		Body:   map[string]any{"email": email},
	}, nil)
}

// ConfirmPasswordReset confirms a password reset token and sets a new
// password.
func (s *RecordService) ConfirmPasswordReset(token, password, passwordConfirm string) error {
	return s.client.Send(s.authPath("confirm-password-reset"), SendOptions{
		Method: "POST",
		Body:   map[string]any{"token": token, "password": password, "passwordConfirm": passwordConfirm},
	}, nil)
}

// RequestEmailChange requires the current session's record to be
// authenticated. Sends a confirmation link to newEmail — the identity
// only changes once that link is confirmed, proving ownership of the new
// address.
func (s *RecordService) RequestEmailChange(newEmail string) error {
	return s.client.Send(s.authPath("request-email-change"), SendOptions{
		Method: "POST",
		Body:   map[string]any{"newEmail": newEmail},
	}, nil)
}

// ConfirmEmailChange confirms an email-change token from the emailed link.
func (s *RecordService) ConfirmEmailChange(token string) error {
	return s.client.Send(s.authPath("confirm-email-change"), SendOptions{
		Method: "POST",
		Body:   map[string]any{"token": token},
	}, nil)
}

File: sdk/dart/lib/src/record_service.dart
import 'client.dart';
import 'types.dart';

/// CRUD + auth operations scoped to one collection. Auth-only methods
/// ([authWithPassword], [authRefresh], ...) only make sense for `auth`-typed
/// collections but are exposed here unconditionally since the client has no
/// way to know a collection's type ahead of a request.
class RecordService {
  final CratebaseClient client;
  final String collectionIdOrName;

  RecordService(this.client, this.collectionIdOrName);

  String get _basePath => '/api/collections/$collectionIdOrName/records';

  Future<ListResult<RecordModel>> getList({
    int page = 1,
    int perPage = 30,
    String? filter,
    String? sort,
  }) async {
    final json = await client.send(
      _basePath,
      query: {'page': page, 'perPage': perPage, 'filter': filter, 'sort': sort},
    );
    return ListResult.fromJson(json as Map<String, dynamic>, RecordModel.fromJson);
  }

  /// Fetches every page and concatenates the results. Convenient for small
  /// collections; prefer [getList] with pagination for large ones.
  Future<List<RecordModel>> getFullList({
    String? filter,
    String? sort,
    int batchSize = 200,
  }) async {
    final items = <RecordModel>[];
    var page = 1;
    for (;;) {
      final result = await getList(page: page, perPage: batchSize, filter: filter, sort: sort);
      items.addAll(result.items);
      if (page >= result.totalPages) break;
      page += 1;
    }
    return items;
  }

  Future<RecordModel> getOne(String id) async {
    final json = await client.send('$_basePath/$id');
    return RecordModel.fromJson(json as Map<String, dynamic>);
  }

  /// [data] is either a `Map<String, dynamic>` (JSON body) or a
  /// [CratebaseMultipart] for collections with file fields.
  Future<RecordModel> create(Object data) async {
    final json = await client.send(_basePath, method: 'POST', body: data);
    return RecordModel.fromJson(json as Map<String, dynamic>);
  }

  /// [data] is either a `Map<String, dynamic>` (JSON body) or a
  /// [CratebaseMultipart] for collections with file fields.
  Future<RecordModel> update(String id, Object data) async {
    final json = await client.send('$_basePath/$id', method: 'PATCH', body: data);
    return RecordModel.fromJson(json as Map<String, dynamic>);
  }

  Future<void> delete(String id) async {
    await client.send('$_basePath/$id', method: 'DELETE');
  }

  /// [identity] is whatever the collection's configured identity field is
  /// (email by default, but could be a username).
  Future<AuthResponse> authWithPassword(String identity, String password) async {
    final json = await client.send(
      '/api/collections/$collectionIdOrName/auth-with-password',
      method: 'POST',
      body: {'identity': identity, 'password': password},
    );
    final result = AuthResponse.fromJson(json as Map<String, dynamic>);
    client.authStore.save(result.token, result.record.toJson());
    return result;
  }

  Future<AuthResponse> authRefresh() async {
    final json = await client.send(
      '/api/collections/$collectionIdOrName/auth-refresh',
      method: 'POST',
    );
    final result = AuthResponse.fromJson(json as Map<String, dynamic>);
    client.authStore.save(result.token, result.record.toJson());
    return result;
  }

  /// Lists available auth methods. Pass your app's OAuth2 redirect
  /// URI/deep link and each provider's `authUrl` comes back ready to open
  /// directly.
  Future<AuthMethodsResponse> listAuthMethods({String? redirectUri}) async {
    final json = await client.send(
      '/api/collections/$collectionIdOrName/auth-methods',
      query: {'redirectUri': redirectUri},
    );
    return AuthMethodsResponse.fromJson(json as Map<String, dynamic>);
  }

  /// Completes an OAuth2 login: exchange the [code] your app received at
  /// [redirectUri] (the same one used to build the `authUrl` from
  /// [listAuthMethods]) for a session.
  Future<AuthResponse> authWithOAuth2(String provider, String code, String redirectUri) async {
    final json = await client.send(
      '/api/collections/$collectionIdOrName/auth-with-oauth2',
      method: 'POST',
      body: {'provider': provider, 'code': code, 'redirectUri': redirectUri},
    );
    final result = AuthResponse.fromJson(json as Map<String, dynamic>);
    client.authStore.save(result.token, result.record.toJson());
    return result;
  }

  /// Sends a verification email if [email] matches an account — always
  /// resolves regardless, so it can't be used to enumerate accounts.
  Future<void> requestVerification(String email) async {
    await client.send(
      '/api/collections/$collectionIdOrName/request-verification',
      method: 'POST',
      body: {'email': email},
    );
  }

  /// Confirms a verification token from the emailed link.
  Future<void> confirmVerification(String token) async {
    await client.send(
      '/api/collections/$collectionIdOrName/confirm-verification',
      method: 'POST',
      body: {'token': token},
    );
  }

  /// Sends a password reset email if [email] matches an account — always
  /// resolves regardless, so it can't be used to enumerate accounts.
  Future<void> requestPasswordReset(String email) async {
    await client.send(
      '/api/collections/$collectionIdOrName/request-password-reset',
      method: 'POST',
      body: {'email': email},
    );
  }

  /// Confirms a password reset token and sets a new password.
  Future<void> confirmPasswordReset(String token, String password, String passwordConfirm) async {
    await client.send(
      '/api/collections/$collectionIdOrName/confirm-password-reset',
      method: 'POST',
      body: {'token': token, 'password': password, 'passwordConfirm': passwordConfirm},
    );
  }

  /// Requires the current session's record to be authenticated. Sends a
  /// confirmation link to [newEmail] — the identity only changes once that
  /// link is confirmed, proving ownership of the new address.
  Future<void> requestEmailChange(String newEmail) async {
    await client.send(
      '/api/collections/$collectionIdOrName/request-email-change',
      method: 'POST',
      body: {'newEmail': newEmail},
    );
  }

  /// Confirms an email-change token from the emailed link.
  Future<void> confirmEmailChange(String token) async {
    await client.send(
      '/api/collections/$collectionIdOrName/confirm-email-change',
      method: 'POST',
      body: {'token': token},
    );
  }
}
